fabtests/hpcs/src/core/core.c

```diff
@@ -0,0 +1,1476 @@
+	if (ret) {
+		hpcs_error("bind_endpoint_completion failed\n");
+		goto err;
+	}
+
+	ret = fi_av_open(domain_state->domain, &av_attr, &domain_state->av, NULL);
+	if (ret) {
+		hpcs_error("unable to open address vector\n");
+		goto err;
+	}
+
+	ret = fi_ep_bind(domain_state->endpoint, &domain_state->av->fid, 0);
+
+	ret = fi_enable(domain_state->endpoint);
+	if (ret) {
+		hpcs_error("error enabling endpoint\n");
+		goto err;
+	}
+
+	ret = fi_getname(&domain_state->endpoint->fid, &our_name, &len);
+
+	if (ret) {
+		hpcs_error("error determining local endpoint name\n");
+		goto err;
+	}
+
+	names = malloc(len * job->ranks);
+	if (names == NULL) {
+		hpcs_error("error allocating memory for address exchange\n");
+		ret = -1;
+		goto err;
+	}
+
+	ret = job->address_exchange(&our_name, names, len, job->ranks);
+	if (ret) {
+		hpcs_error("error exchanging addresses\n");
+		goto err;
+	}
+
+	ret = fi_av_insert(domain_state->av, names, job->ranks,
+			   domain_state->addresses, 0, NULL);
+	if (ret != job->ranks) {
+		hpcs_error("unable to insert all addresses into AV table\n");
+		ret = -1;
+		goto err;
+	} else {
+		ret = 0;
+	}
+
+	if(verbose){
+		hpcs_verbose("Rank %d peer addresses: ", job->rank);
+		for (i = 0; i < job->ranks; i++)
+			printf("%d:%lx ", i, (uint64_t)(domain_state->addresses[i]));
+		printf("\n");
+	}
+
+
+err:
+	free (names);
+	return ret;
+}
+
+/*
+ * Initializes OFI resources.
+ */
+static int init_ofi(
+		const struct arguments *arguments,
+		const void *test_arguments,
+		const struct test_config *test_config,
+		struct ofi_state *ofi_state,
+		struct job *job)
+{
+	int ret;
+	struct fi_info *hints;
+	struct fi_info *info;
+	const char *node = NULL;
+	const char *service = "2042";
+	const uint64_t flags = 0;
+	void *context = NULL;
+
+	hints = fi_allocinfo();
+	if (!hints)
+		return -FI_ENOMEM;
+
+	hints->fabric_attr->prov_name = arguments->prov_name;
+	hints->domain_attr->resource_mgmt = FI_RM_ENABLED;
+	hints->domain_attr->mr_mode = FI_MR_RMA_EVENT;
+	hints->domain_attr->mr_key_size = 4;
+	hints->caps = test_config->minimum_caps |
+			(arguments->pattern_api.enable_triggered ? FI_TRIGGER : 0);
+	hints->mode = FI_CONTEXT2;
+	hints->ep_attr->type = FI_EP_RDM;
+	hints->tx_attr->op_flags = FI_DELIVERY_COMPLETE;
+
+	ret = fi_getinfo(FI_VERSION(1,6), node, service, flags, hints, &info);
+	if (ret)
+		goto err_getinfo;
+
+	ret = fi_fabric(info->fabric_attr, &ofi_state->fabric, context);
+	if (ret) {
+		hpcs_error("fi_fabric failed\n");
+		goto err_fabric;
+	}
+
+	ret = init_domain(arguments, test_config, ofi_state->fabric, info,
+			ofi_state->domain_state, job);
+	if (ret) {
+		hpcs_error("init_domain failed\n");
+		goto err_domain;
+	} else {
+		hpcs_verbose("init_domain complete, Using %s provider\n",
+			     info->fabric_attr->name);
+	}
+
+	return 0;
+
+err_domain:
+	fi_close(&ofi_state->fabric->fid);
+
+err_fabric:
+	fi_freeinfo(info);
+
+err_getinfo:
+	fi_freeinfo(hints);
+
+	return ret;
+}
```

Comment:
It would be ideal to re-use the code in common/shared.c.  The shared code would likely need to be restructured to support this.